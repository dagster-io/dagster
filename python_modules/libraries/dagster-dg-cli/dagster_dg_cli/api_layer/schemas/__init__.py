"""Pydantic schemas for Dagster Plus API."""
