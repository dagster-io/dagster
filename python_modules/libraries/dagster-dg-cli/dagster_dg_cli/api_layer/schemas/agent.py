"""Agent models for REST-like API."""

from enum import Enum
from typing import Optional

from pydantic import BaseModel


class DgApiAgentStatus(str, Enum):
    """Agent status enum for FastAPI compatibility."""

    RUNNING = "RUNNING"
    STOPPED = "STOPPED"
    NOT_RUNNING = "NOT_RUNNING"
    UNHEALTHY = "UNHEALTHY"
    UNKNOWN = "UNKNOWN"


class DgApiAgentMetadataEntry(BaseModel):
    """Agent metadata key-value pair."""

    key: str
    value: str


class DgApiAgent(BaseModel):
    """Agent resource model."""

    id: str  # Agent IDs are strings in the GraphQL schema
    agent_label: Optional[str]  # Can be None in some cases
    status: DgApiAgentStatus
    last_heartbeat_time: Optional[float]
    metadata: list[DgApiAgentMetadataEntry]

    class Config:
        from_attributes = True  # For future ORM compatibility


class DgApiAgentList(BaseModel):
    """GET /api/agents response."""

    items: list[DgApiAgent]
    total: int
