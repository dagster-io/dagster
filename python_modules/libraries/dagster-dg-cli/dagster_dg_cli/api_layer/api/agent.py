"""Agent endpoints - REST-like interface."""

from dataclasses import dataclass
from typing import TYPE_CHECKING, Optional

from dagster_dg_cli.api_layer.graphql_adapter.agent import (
    get_agent_via_graphql,
    list_agents_via_graphql,
)
from dagster_dg_cli.utils.plus.gql_client import IGraphQLClient

if TYPE_CHECKING:
    from dagster_dg_cli.api_layer.schemas.agent import DgApiAgent, DgApiAgentList


@dataclass(frozen=True)
class DgApiAgentApi:
    client: IGraphQLClient

    def list_agents(self) -> "DgApiAgentList":
        return list_agents_via_graphql(self.client)

    def get_agent(self, agent_id: str) -> Optional["DgApiAgent"]:
        return get_agent_via_graphql(self.client, agent_id)
