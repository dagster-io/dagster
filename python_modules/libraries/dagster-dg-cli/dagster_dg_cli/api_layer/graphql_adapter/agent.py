"""GraphQL implementation for agent operations."""

from typing import TYPE_CHECKING, Any, Optional

from dagster_dg_cli.utils.plus.gql_client import IGraphQLClient

if TYPE_CHECKING:
    from dagster_dg_cli.api_layer.schemas.agent import DgApiAgent, DgApiAgentList

# GraphQL queries
LIST_AGENTS_QUERY = """
query ListAgents {
    agents {
        id
        agentLabel
        status
        lastHeartbeatTime
        metadata {
            key
            value
        }
    }
}
"""

# Note: There's no single 'agent' query, so we use 'agents' and filter client-side


def process_agents_response(graphql_response: dict[str, Any]) -> "DgApiAgentList":
    """Process GraphQL response into AgentList.

    This is a pure function that can be easily tested without mocking GraphQL clients.

    Args:
        graphql_response: Raw GraphQL response containing "agents"

    Returns:
        AgentList: Processed agent data
    """
    # Import pydantic models only when needed
    from dagster_dg_cli.api_layer.schemas.agent import (
        DgApiAgent,
        DgApiAgentList,
        DgApiAgentMetadataEntry,
        DgApiAgentStatus,
    )

    agents_data = graphql_response.get("agents", [])

    agents = []
    for a in agents_data:
        metadata = [
            DgApiAgentMetadataEntry(key=m["key"], value=m["value"]) for m in a.get("metadata", [])
        ]

        agents.append(
            DgApiAgent(
                id=a["id"],
                agent_label=a.get("agentLabel"),
                status=DgApiAgentStatus[a["status"]],
                last_heartbeat_time=a.get("lastHeartbeatTime"),
                metadata=metadata,
            )
        )

    return DgApiAgentList(
        items=agents,
        total=len(agents),
    )


def list_agents_via_graphql(
    client: IGraphQLClient,
    limit: Optional[int] = None,
) -> "DgApiAgentList":
    """Fetch agents using GraphQL.
    This is an implementation detail that can be replaced with REST calls later.
    """
    result = client.execute(LIST_AGENTS_QUERY)

    agent_list = process_agents_response(result)

    # Apply limit if specified
    if limit is not None:
        agent_list.items = agent_list.items[:limit]
        agent_list.total = len(agent_list.items)

    return agent_list


def get_agent_via_graphql(
    client: IGraphQLClient,
    agent_id: str,
) -> Optional["DgApiAgent"]:
    """Fetch single agent using GraphQL.
    This is an implementation detail that can be replaced with REST calls later.

    Note: Since there's no single 'agent' query, we fetch all agents and filter client-side.
    """
    result = client.execute(LIST_AGENTS_QUERY)
    agent_list = process_agents_response(result)

    # Find the agent with matching ID
    for agent in agent_list.items:
        if agent.id == agent_id:
            return agent

    return None
