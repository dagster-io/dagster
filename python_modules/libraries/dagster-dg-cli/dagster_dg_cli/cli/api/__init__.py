"""API commands for interacting with Dagster Plus REST-like interface."""

from dagster_dg_cli.cli.api.cli_group import api_group

__all__ = ["api_group"]
