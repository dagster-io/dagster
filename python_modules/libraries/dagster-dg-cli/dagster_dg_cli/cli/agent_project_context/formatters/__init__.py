"""Output formatters for project context."""

from abc import ABC, abstractmethod

from dagster_dg_cli.cli.agent_project_context.models import ProjectContext


class BaseFormatter(ABC):
    """Base class for project context formatters."""

    @abstractmethod
    def format(self, context: ProjectContext) -> str:
        """Format project context into string output."""
        pass


def get_formatter(format_name: str) -> BaseFormatter:
    """Get formatter instance by name."""
    from dagster_dg_cli.cli.agent_project_context.formatters.json import JsonFormatter
    from dagster_dg_cli.cli.agent_project_context.formatters.markdown import MarkdownFormatter
    from dagster_dg_cli.cli.agent_project_context.formatters.xml import XmlFormatter

    formatters = {
        "json": JsonFormatter,
        "markdown": MarkdownFormatter,
        "xml": XmlFormatter,
    }

    formatter_class = formatters.get(format_name)
    if not formatter_class:
        raise ValueError(f"Unknown format: {format_name}")

    return formatter_class()
