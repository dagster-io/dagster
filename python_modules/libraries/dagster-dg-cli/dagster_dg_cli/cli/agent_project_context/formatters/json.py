"""JSON formatter for project context."""

import json

from dagster_dg_cli.cli.agent_project_context.formatters import BaseFormatter
from dagster_dg_cli.cli.agent_project_context.models import ProjectContext


class JsonFormatter(BaseFormatter):
    """Formats project context as JSON."""

    def format(self, context: ProjectContext) -> str:
        """Format project context as JSON."""
        # Convert record objects to dictionaries for JSON serialization
        context_dict = {
            "working_directory": context.working_directory,
            "project_root": context.project_root,
            "project_name": context.project_name,
            "components": [
                {
                    "key": comp.key,
                    "summary": comp.summary,
                    "namespace": comp.namespace,
                    "name": comp.name,
                }
                for comp in context.components
            ],
            "integrations": [
                {
                    "name": integ.name,
                    "description": integ.description,
                    "pypi": integ.pypi,
                }
                for integ in context.integrations
            ],
            "project_structure": {
                "has_pyproject_toml": context.project_structure.has_pyproject_toml,
                "has_dbt_project": context.project_structure.has_dbt_project,
                "dbt_project_paths": context.project_structure.dbt_project_paths,
                "python_packages": context.project_structure.python_packages,
            },
            "dbt_schemas": {
                key: {
                    "description": schema.description,
                    "scaffold_params_schema": schema.scaffold_params_schema,
                    "component_schema": schema.component_schema,
                }
                for key, schema in context.dbt_schemas.items()
            },
        }
        return json.dumps(context_dict, indent=2)
