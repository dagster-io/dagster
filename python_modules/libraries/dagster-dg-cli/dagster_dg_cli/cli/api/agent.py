"""Agent API commands following GitHub CLI patterns."""

import json

import click
from dagster_dg_core.utils import DgClickCommand, DgClickGroup
from dagster_dg_core.utils.telemetry import cli_telemetry_wrapper
from dagster_shared.plus.config import DagsterPlusCliConfig
from dagster_shared.plus.config_utils import dg_api_options

# Lazy import to avoid loading pydantic at CLI startup
from dagster_dg_cli.cli.api.client import create_dg_api_graphql_client
from dagster_dg_cli.cli.api.formatters import format_agent, format_agents


@click.command(name="list", cls=DgClickCommand, unlaunched=True)
@click.option(
    "--json",
    "output_json",
    is_flag=True,
    help="Output in JSON format for machine readability",
)
@dg_api_options(organization_scoped=True)
@cli_telemetry_wrapper
@click.pass_context
def list_agents_command(
    ctx: click.Context, output_json: bool, organization: str, api_token: str, view_graphql: bool
) -> None:
    """List all agents in the organization."""
    config = DagsterPlusCliConfig.create_for_organization(
        organization=organization,
        user_token=api_token,
    )
    client = create_dg_api_graphql_client(ctx, config, view_graphql=view_graphql)
    from dagster_dg_cli.api_layer.api.agent import DgApiAgentApi

    api = DgApiAgentApi(client)

    try:
        agents = api.list_agents()
        output = format_agents(agents, as_json=output_json)
        click.echo(output)
    except Exception as e:
        if output_json:
            error_response = {"error": str(e)}
            click.echo(json.dumps(error_response), err=True)
        else:
            click.echo(f"Error querying Dagster Plus API: {e}", err=True)
        raise click.ClickException(f"Failed to list agents: {e}")


@click.command(name="get", cls=DgClickCommand, unlaunched=True)
@click.argument("agent_id", type=str)
@click.option(
    "--json",
    "output_json",
    is_flag=True,
    help="Output in JSON format for machine readability",
)
@dg_api_options(organization_scoped=True)
@cli_telemetry_wrapper
@click.pass_context
def get_agent_command(
    ctx: click.Context,
    agent_id: str,
    output_json: bool,
    organization: str,
    api_token: str,
    view_graphql: bool,
) -> None:
    """Get detailed information about a specific agent."""
    config = DagsterPlusCliConfig.create_for_organization(
        organization=organization,
        user_token=api_token,
    )
    client = create_dg_api_graphql_client(ctx, config, view_graphql=view_graphql)
    from dagster_dg_cli.api_layer.api.agent import DgApiAgentApi

    api = DgApiAgentApi(client)

    try:
        agent = api.get_agent(agent_id)
        if agent is None:
            if output_json:
                error_response = {"error": f"Agent with ID '{agent_id}' not found"}
                click.echo(json.dumps(error_response), err=True)
            else:
                click.echo(f"Agent with ID '{agent_id}' not found", err=True)
            raise click.ClickException(f"Agent not found: {agent_id}")

        output = format_agent(agent, as_json=output_json)
        click.echo(output)
    except Exception as e:
        if output_json:
            error_response = {"error": str(e)}
            click.echo(json.dumps(error_response), err=True)
        else:
            click.echo(f"Error querying Dagster Plus API: {e}", err=True)
        raise click.ClickException(f"Failed to get agent: {e}")


@click.group(
    name="agent",
    cls=DgClickGroup,
    unlaunched=True,
    commands={
        "list": list_agents_command,
        "get": get_agent_command,
    },
)
def agent_group():
    """Manage agents in Dagster Plus."""
