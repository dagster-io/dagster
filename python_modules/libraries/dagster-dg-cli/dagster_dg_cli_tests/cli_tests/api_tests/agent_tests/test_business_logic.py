"""Test agent business logic functions without mocks.

These tests focus on testing pure functions that process data without requiring
GraphQL client mocking or external dependencies.
"""

from dagster_dg_cli.api_layer.schemas.agent import (
    DgApiAgent,
    DgApiAgentList,
    DgApiAgentMetadataEntry,
    DgApiAgentStatus,
)
from dagster_dg_cli.cli.api.formatters import format_agent, format_agents


class TestFormatAgents:
    """Test the agent formatting functions."""

    def _create_sample_agent_list(self):
        """Create sample AgentList for testing."""
        agents = [
            DgApiAgent(
                id="agent-1-uuid-12345",
                agent_label="Production Agent",
                status=DgApiAgentStatus.RUNNING,
                last_heartbeat_time=1641046800.0,  # 2022-01-01 14:20:00 UTC (midday to avoid timezone edge cases)
                metadata=[
                    DgApiAgentMetadataEntry(key="version", value="1.2.3"),
                    DgApiAgentMetadataEntry(key="location", value="us-east-1"),
                ],
            ),
            DgApiAgent(
                id="agent-2-uuid-67890",
                agent_label=None,  # No label - should fall back to ID display
                status=DgApiAgentStatus.STOPPED,
                last_heartbeat_time=None,
                metadata=[],
            ),
            DgApiAgent(
                id="agent-3-uuid-abcdef",
                agent_label="Staging Agent",
                status=DgApiAgentStatus.UNHEALTHY,
                last_heartbeat_time=1641046860.0,  # 2022-01-01 14:21:00 UTC
                metadata=[
                    DgApiAgentMetadataEntry(key="environment", value="staging"),
                ],
            ),
        ]
        return DgApiAgentList(items=agents, total=3)

    def _create_empty_agent_list(self):
        """Create empty AgentList for testing."""
        return DgApiAgentList(items=[], total=0)

    def _create_single_agent(self):
        """Create single Agent for testing."""
        return DgApiAgent(
            id="single-agent-uuid-xyz",
            agent_label="Development Agent",
            status=DgApiAgentStatus.RUNNING,
            last_heartbeat_time=1641046800.0,
            metadata=[
                DgApiAgentMetadataEntry(key="owner", value="dev-team"),
                DgApiAgentMetadataEntry(key="cpu_limit", value="2"),
                DgApiAgentMetadataEntry(key="memory_limit", value="4Gi"),
            ],
        )

    def test_format_agents_text_output(self, snapshot):
        """Test formatting agents as text."""
        from dagster_shared.utils.timing import fixed_timezone

        agent_list = self._create_sample_agent_list()
        with fixed_timezone("UTC"):
            result = format_agents(agent_list, as_json=False)

        # Snapshot the entire text output
        snapshot.assert_match(result)

    def test_format_agents_json_output(self, snapshot):
        """Test formatting agents as JSON."""
        agent_list = self._create_sample_agent_list()
        result = format_agents(agent_list, as_json=True)

        # For JSON, we want to snapshot the parsed structure to avoid formatting differences
        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_empty_agents_text_output(self, snapshot):
        """Test formatting empty agent list as text."""
        agent_list = self._create_empty_agent_list()
        result = format_agents(agent_list, as_json=False)

        snapshot.assert_match(result)

    def test_format_empty_agents_json_output(self, snapshot):
        """Test formatting empty agent list as JSON."""
        agent_list = self._create_empty_agent_list()
        result = format_agents(agent_list, as_json=True)

        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_single_agent_text_output(self, snapshot):
        """Test formatting single agent as text."""
        from dagster_shared.utils.timing import fixed_timezone

        agent = self._create_single_agent()
        with fixed_timezone("UTC"):
            result = format_agent(agent, as_json=False)

        snapshot.assert_match(result)

    def test_format_single_agent_json_output(self, snapshot):
        """Test formatting single agent as JSON."""
        agent = self._create_single_agent()
        result = format_agent(agent, as_json=True)

        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_agent_without_metadata(self, snapshot):
        """Test formatting agent with no metadata."""
        from dagster_shared.utils.timing import fixed_timezone

        agent = DgApiAgent(
            id="simple-agent-uuid",
            agent_label="Simple Agent",
            status=DgApiAgentStatus.NOT_RUNNING,
            last_heartbeat_time=None,
            metadata=[],
        )
        with fixed_timezone("UTC"):
            result = format_agent(agent, as_json=False)

        snapshot.assert_match(result)

    def test_format_agent_without_label(self, snapshot):
        """Test formatting agent with no label (should use ID fallback)."""
        from dagster_shared.utils.timing import fixed_timezone

        agent = DgApiAgent(
            id="no-label-agent-uuid-123456789",
            agent_label=None,
            status=DgApiAgentStatus.UNKNOWN,
            last_heartbeat_time=1641046920.0,  # 2022-01-01 14:22:00 UTC
            metadata=[
                DgApiAgentMetadataEntry(key="type", value="serverless"),
            ],
        )
        with fixed_timezone("UTC"):
            result = format_agent(agent, as_json=False)

        snapshot.assert_match(result)


class TestAgentDataProcessing:
    """Test processing of agent data structures.

    This class would test any pure functions in the GraphQL adapter
    that process the raw GraphQL responses into our domain models.
    Since the actual GraphQL processing is done inline in the adapter
    functions, these tests will verify our data model creation.
    """

    def test_agent_creation_with_all_statuses(self, snapshot):
        """Test creating agents with all possible status values."""
        agents = [
            DgApiAgent(
                id=f"agent-{status.value.lower()}-uuid",
                agent_label=f"Agent {status.value.title()}",
                status=status,
                last_heartbeat_time=1641046800.0 if status == DgApiAgentStatus.RUNNING else None,
                metadata=[
                    DgApiAgentMetadataEntry(key="status_test", value=status.value),
                ],
            )
            for status in DgApiAgentStatus
        ]

        agent_list = DgApiAgentList(items=agents, total=len(agents))

        # Test JSON serialization works correctly for all statuses
        result = agent_list.model_dump_json(indent=2)
        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_agent_metadata_handling(self):
        """Test agent metadata entry creation and access."""
        agent = DgApiAgent(
            id="metadata-test-agent",
            agent_label="Metadata Test",
            status=DgApiAgentStatus.RUNNING,
            last_heartbeat_time=1641046800.0,
            metadata=[
                DgApiAgentMetadataEntry(key="version", value="1.0.0"),
                DgApiAgentMetadataEntry(key="environment", value="production"),
                DgApiAgentMetadataEntry(key="region", value="us-west-2"),
            ],
        )

        assert len(agent.metadata) == 3
        assert agent.metadata[0].key == "version"
        assert agent.metadata[0].value == "1.0.0"
        assert agent.metadata[1].key == "environment"
        assert agent.metadata[1].value == "production"
        assert agent.metadata[2].key == "region"
        assert agent.metadata[2].value == "us-west-2"

    def test_agent_list_total_count(self):
        """Test that AgentList properly tracks total count."""
        agents = [
            DgApiAgent(
                id=f"agent-{i}",
                agent_label=f"Agent {i}",
                status=DgApiAgentStatus.RUNNING,
                last_heartbeat_time=None,
                metadata=[],
            )
            for i in range(3)
        ]

        agent_list = DgApiAgentList(
            items=agents, total=10
        )  # Total could be different from items length (pagination)

        assert len(agent_list.items) == 3
        assert agent_list.total == 10

    def test_agent_id_fallback_display(self):
        """Test agent display label fallback behavior."""
        # Test with label
        agent_with_label = DgApiAgent(
            id="very-long-agent-uuid-12345678901234567890",
            agent_label="Custom Label",
            status=DgApiAgentStatus.RUNNING,
            last_heartbeat_time=None,
            metadata=[],
        )

        # Test without label
        agent_without_label = DgApiAgent(
            id="very-long-agent-uuid-12345678901234567890",
            agent_label=None,
            status=DgApiAgentStatus.RUNNING,
            last_heartbeat_time=None,
            metadata=[],
        )

        # Format both agents and check the label display
        result_with_label = format_agent(agent_with_label, as_json=False)
        result_without_label = format_agent(agent_without_label, as_json=False)

        assert "Label: Custom Label" in result_with_label
        assert "Label: Agent very-lon" in result_without_label  # Should show first 8 chars
