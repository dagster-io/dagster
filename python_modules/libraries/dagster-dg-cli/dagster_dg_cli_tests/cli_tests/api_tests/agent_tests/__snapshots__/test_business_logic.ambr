# serializer version: 1
# name: TestAgentDataProcessing.test_agent_creation_with_all_statuses[0]
  dict({
    'items': list([
      dict({
        'agent_label': 'Agent Running',
        'id': 'agent-running-uuid',
        'last_heartbeat_time': 1641046800.0,
        'metadata': list([
          dict({
            'key': 'status_test',
            'value': 'RUNNING',
          }),
        ]),
        'status': 'RUNNING',
      }),
      dict({
        'agent_label': 'Agent Stopped',
        'id': 'agent-stopped-uuid',
        'last_heartbeat_time': None,
        'metadata': list([
          dict({
            'key': 'status_test',
            'value': 'STOPPED',
          }),
        ]),
        'status': 'STOPPED',
      }),
      dict({
        'agent_label': 'Agent Not_Running',
        'id': 'agent-not_running-uuid',
        'last_heartbeat_time': None,
        'metadata': list([
          dict({
            'key': 'status_test',
            'value': 'NOT_RUNNING',
          }),
        ]),
        'status': 'NOT_RUNNING',
      }),
      dict({
        'agent_label': 'Agent Unhealthy',
        'id': 'agent-unhealthy-uuid',
        'last_heartbeat_time': None,
        'metadata': list([
          dict({
            'key': 'status_test',
            'value': 'UNHEALTHY',
          }),
        ]),
        'status': 'UNHEALTHY',
      }),
      dict({
        'agent_label': 'Agent Unknown',
        'id': 'agent-unknown-uuid',
        'last_heartbeat_time': None,
        'metadata': list([
          dict({
            'key': 'status_test',
            'value': 'UNKNOWN',
          }),
        ]),
        'status': 'UNKNOWN',
      }),
    ]),
    'total': 5,
  })
# ---
# name: TestFormatAgents.test_format_agent_without_label[0]
  '''
  Label: Agent no-label
  ID: no-label-agent-uuid-123456789
  Status: UNKNOWN
  Last Heartbeat: 2022-01-01 14:22:00
  
  Metadata:
    type: serverless
  '''
# ---
# name: TestFormatAgents.test_format_agent_without_metadata[0]
  '''
  Label: Simple Agent
  ID: simple-agent-uuid
  Status: NOT_RUNNING
  Last Heartbeat: Never
  '''
# ---
# name: TestFormatAgents.test_format_agents_json_output[0]
  dict({
    'items': list([
      dict({
        'agent_label': 'Production Agent',
        'id': 'agent-1-uuid-12345',
        'last_heartbeat_time': 1641046800.0,
        'metadata': list([
          dict({
            'key': 'version',
            'value': '1.2.3',
          }),
          dict({
            'key': 'location',
            'value': 'us-east-1',
          }),
        ]),
        'status': 'RUNNING',
      }),
      dict({
        'agent_label': None,
        'id': 'agent-2-uuid-67890',
        'last_heartbeat_time': None,
        'metadata': list([
        ]),
        'status': 'STOPPED',
      }),
      dict({
        'agent_label': 'Staging Agent',
        'id': 'agent-3-uuid-abcdef',
        'last_heartbeat_time': 1641046860.0,
        'metadata': list([
          dict({
            'key': 'environment',
            'value': 'staging',
          }),
        ]),
        'status': 'UNHEALTHY',
      }),
    ]),
    'total': 3,
  })
# ---
# name: TestFormatAgents.test_format_agents_text_output[0]
  '''
  Label: Production Agent
  ID: agent-1-uuid-12345
  Status: RUNNING
  Last Heartbeat: 2022-01-01 14:20:00
  
  Label: Agent agent-2-
  ID: agent-2-uuid-67890
  Status: STOPPED
  Last Heartbeat: Never
  
  Label: Staging Agent
  ID: agent-3-uuid-abcdef
  Status: UNHEALTHY
  Last Heartbeat: 2022-01-01 14:21:00
  '''
# ---
# name: TestFormatAgents.test_format_empty_agents_json_output[0]
  dict({
    'items': list([
    ]),
    'total': 0,
  })
# ---
# name: TestFormatAgents.test_format_empty_agents_text_output[0]
  ''
# ---
# name: TestFormatAgents.test_format_single_agent_json_output[0]
  dict({
    'agent_label': 'Development Agent',
    'id': 'single-agent-uuid-xyz',
    'last_heartbeat_time': 1641046800.0,
    'metadata': list([
      dict({
        'key': 'owner',
        'value': 'dev-team',
      }),
      dict({
        'key': 'cpu_limit',
        'value': '2',
      }),
      dict({
        'key': 'memory_limit',
        'value': '4Gi',
      }),
    ]),
    'status': 'RUNNING',
  })
# ---
# name: TestFormatAgents.test_format_single_agent_text_output[0]
  '''
  Label: Development Agent
  ID: single-agent-uuid-xyz
  Status: RUNNING
  Last Heartbeat: 2022-01-01 14:20:00
  
  Metadata:
    owner: dev-team
    cpu_limit: 2
    memory_limit: 4Gi
  '''
# ---
